### ISSUE TITLE
New analyser: better processing of import cycles with star imports

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#7141]
Consider this test case:
```
[case testClassDeferredStar]
import lib
[file lib.py]
from other import *

class C:
    x: int
[file other.py]
import lib
```
Currently the order of processing targets is like this:
* `lib`
* `other`
* `lib`
* `__main__`

I think it should be possible to process such SCC without deferral at all. Couple ideas to achieve this:
* Better ordering of targets within SCC
* Make `missing_names` act per namespace (to avoid extra deferral in more complex scenarios)

**Comment by ilevkivskyi** [comment]
Currently this affects two plugin test cases `testTypeInAttrDeferredStar` and `testTypeInDataclassDeferredStar`.

**Comment by JukkaL** [comment]
It looks like we generally process modules in a very sub-optimal order. I'll see what happens if we reverse the order.

**Comment by JukkaL** [comment]
Fixed by #7163.
